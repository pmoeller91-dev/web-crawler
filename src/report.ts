import { PagesMap } from "./crawl.js";

/**
 * Takes a map of pages and number of occurrences, and returns a formatted
 * report as a string to print to the console.
 * @param pages - A map containing normalized page URLs and the number of
 * occurrences of each.
 * @returns - A string of the entire report, ready to be output.
 */
const printReport = (pages: PagesMap): string => {
  const lines: string[] = [];
  // Sort ascending by number of occurrences
  const sortedPages = [...pages].sort((a, b) => b[1] - a[1]);
  lines.push("Starting pages report...");
  lines.push("");
  lines.push(
    ...sortedPages.map(
      ([page, numOccurrences]) =>
        `Found ${numOccurrences} internal link${numOccurrences === 1 ? "" : "s"} to "${page}"`,
    ),
  );
  const totalPages = sortedPages.length;
  const totalLinks = sortedPages.reduce((a, x) => a + x[1], 0);
  lines.push("");
  lines.push(
    `Pages report completed. Crawled ${totalPages} page${totalPages === 1 ? "" : "s"} and found ${totalLinks} internal link${totalLinks === 1 ? "" : "s"}.`,
  );
  return lines.join("\n");
};

export { printReport };
