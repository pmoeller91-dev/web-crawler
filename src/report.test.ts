import { describe, it, expect } from "@jest/globals";
import { printReport } from "./report.js";
import type { PagesMap } from "./crawl.js";

describe("printReport", () => {
  it("should print a nicely formatted report with pages in descending order", () => {
    const pages: PagesMap = new Map();
    pages.set("yahoo.com", 10);
    pages.set("google.com", 100);
    pages.set("bing.com", 50);
    const report = printReport(pages);
    expect(report).toMatchSnapshot();
    expect(report.indexOf("google.com")).toBeLessThan(report.indexOf("bing.com"));
    expect(report.indexOf("bing.com")).toBeLessThan(report.indexOf("yahoo.com"));
  });
});
