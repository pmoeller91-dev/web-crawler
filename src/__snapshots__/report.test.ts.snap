// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`printReport should print a nicely formatted report with pages in descending order 1`] = `
"Starting pages report...

Found 100 internal links to "google.com"
Found 50 internal links to "bing.com"
Found 10 internal links to "yahoo.com"

Pages report completed. Crawled 3 pages and found 160 internal links."
`;
